lnwire/custom.go

```diff
@@ -2,14 +2,71 @@ package lnwire
 
 import (
 	"bytes"
-	"errors"
+	"fmt"
 	"io"
+	"sync"
 )
 
 // CustomTypeStart is the start of the custom type range for peer messages as
 // defined in BOLT 01.
 const CustomTypeStart MessageType = 32768
 
+var (
+	// customTypeOverride contains a set of message types < CustomTypeStart
+	// that lnd allows to be treated as custom messages. This allows us to
+	// override messages reserved for the protocol level and treat them as
+	// custom messages. This set of message types is stored as a global so
+	// that we do not need to pass around state when accounting for this
+	// set of messages in message creation.
+	//
+	// Note: This global is protected by the customTypeOverride mutex.
+	customTypeOverride map[MessageType]struct{}
+
+	// customTypeOverrideMtx manages concurrent access to
+	// customTypeOverride.
+	customTypeOverrideMtx sync.RWMutex
+)
+
+// SetCustomOverrides validates that the set of override types are outside of
+// the custom message range (there's no reason to override messages that are
+// already within the range), and updates the customTypeOverride global to hold
+// this set of message types.
+func SetCustomOverrides(overrideTypes []uint16) error {
+	customTypeOverrideMtx.Lock()
+	defer customTypeOverrideMtx.Unlock()
+
+	customTypeOverride = make(map[MessageType]struct{}, len(overrideTypes))
+
+	for _, t := range overrideTypes {
+		msgType := MessageType(t)
+
+		if msgType >= CustomTypeStart {
+			return fmt.Errorf("can't override type: %v, already "+
+				"in custom range", t)
+		}
+
+		customTypeOverride[msgType] = struct{}{}
+	}
+
+	return nil
+}
+
+// IsCustomOverride returns a bool indicating whether the message type is one
+// of the protocol messages that we override for custom use.
+func IsCustomOverride(t MessageType) bool {
+	customTypeOverrideMtx.RLock()
+	defer customTypeOverrideMtx.RUnlock()
+
+	// If our map is nil, we've never set any custom overrides so the
+	// message can't possibly be overridden.
+	if customTypeOverride == nil {
```

Comment:
I don't think this can be hit since `SetCustomOverrides` is always called and will initialize `customTypeOverride`

Reply:
I left this in to safeguard against a panic if we call `IsCustomOverride` before `SetCustomOverrides`, but I guess that's desired behavior because we should never call the check before the we set (which happens in config parsing, so really should never happen)?